fluent-syntax/src/parser/mod.rs

```diff
@@ -1,14 +1,842 @@
+                true
+            } else {
+                false
+            }
+        }
+        ast::InlineExpression::CallExpression { ref callee, .. } => {
+            if let ast::InlineExpression::FunctionReference { .. } = **callee {
+                true
+            } else if let ast::InlineExpression::AttributeExpression { .. } = **callee {
+                true
+            } else {
+                false
+            }
+        }
+        _ => false,
+    };
+
+    if !is_valid {
+        //XXX: Generalize error type
+        return error!(ErrorKind::MessageReferenceAsSelector, ps.ptr);
+    }
+
+    ps.ptr += 2; // ->
+
+    ps.skip_blank_inline();
+    ps.expect_byte(b'\n')?;
+    ps.skip_blank();
+
+    let variants = get_variants(ps)?;
+
+    Ok(ast::Expression::SelectExpression {
+        selector: exp,
+        variants,
+    })
+}
+
+fn get_call_expression<'p>(ps: &mut ParserStream<'p>) -> Result<ast::InlineExpression<'p>> {
+    let mut callee = get_attribute_expression(ps)?;
+
+    let expr = if ps.is_current_byte(b'(') {
+        let is_valid = match callee {
+            ast::InlineExpression::AttributeExpression { ref reference, .. } => {
+                if let ast::InlineExpression::TermReference { .. } = **reference {
+                    true
+                } else {
+                    false
+                }
+            }
+            ast::InlineExpression::TermReference { .. } => true,
+            ast::InlineExpression::MessageReference { ref id, .. } => {
+                id.name.find(|c: char| c.is_ascii_lowercase()).is_none()
+            }
+            _ => false,
+        };
+
+        if is_valid {
+            if let ast::InlineExpression::MessageReference { id } = callee {
+                callee = ast::InlineExpression::FunctionReference { id };
+            }
+            let (positional, named) = get_call_args(ps)?;
+            ast::InlineExpression::CallExpression {
+                callee: Box::new(callee),
+                positional,
+                named,
+            }
+        } else {
+            return error!(ErrorKind::ForbiddenCallee, ps.ptr);
+        }
+    } else {
+        callee
+    };
+
+    Ok(expr)
+}
+
+fn get_attribute_expression<'p>(ps: &mut ParserStream<'p>) -> Result<ast::InlineExpression<'p>> {
+    let reference = get_literal(ps)?;
+
+    match reference {
+        ast::InlineExpression::MessageReference { .. }
+        | ast::InlineExpression::TermReference { .. } => {
+            if ps.is_current_byte(b'.') {
+                ps.ptr += 1; // .
+                let attr = get_identifier(ps)?;
+                Ok(ast::InlineExpression::AttributeExpression {
+                    reference: Box::new(reference),
+                    name: attr,
+                })
+            } else {
+                Ok(reference)
+            }
+        }
+        _ => Ok(reference),
+    }
+}
+
+fn get_literal<'p>(ps: &mut ParserStream<'p>) -> Result<ast::InlineExpression<'p>> {
```

Comment:
There are only two literals in `Fluent` spec: `StringLiteral` and `NumberLiteral`. In `fluent.js` we called this function `getSimpleExpression`.